### PR TITLE
adding agent name feature to help monitor traces in LLMOps tools

### PR DESCRIPTION
**Pull request by saibhaskerraju**
Hello,

I have been using various LLM monitoring tools and I noticed that the Trace Name appears as "CrewAgentExecutor" by default in monitoring tools

below image for reference
image

It can be helpful for developers if we add a name to the Agent so that the traces have these names
Below image for reference

image

This helps in filtering and easy monitoring of traces in any LLMOps tool.

### CODE REVIEW

**Comment by joaomdmoura** [comment]
I can see how useful this would be, can we force the name to be the role value though?

**Comment by saibhaskerraju** [comment]
you are correct @joaomdmoura , I didnt notice that param earlier. thanks.

I changed the code. please review

**Comment by gvieira** [comment]
Nice improvement! Would you mind adding a test (or a few) for that?

**Comment by pythonbyte** [comment]
Hi @saibhaskerraju, thanks for the PR. 

Hey there! Could you take a moment to check if this still fits your needs? If it does, just let us know, and we’ll be happy to help you update and merge it. Also, there are a few conflicts we’ll need to sort out together. Thanks!

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comments for PR #493

## Overview
This pull request introduces critical changes to the `src/crewai/agent.py` file, focusing on enhancing agent naming functionality and integrating decorators for tracking agent behavior. Below are the key findings from the review, along with recommendations for improvements.

### 1. Agent Name Implementation (Patch 1 & Patch 4)

- **Initial Implementation**:
  ```python
  agent_name: Optional[str] = Field(
      default=None, description="The Name of the Agent"
  )
  ```
  The addition of the `agent_name` field was well-intentioned for clarity but was subsequently removed in favor of utilizing the existing `role` field. This change was beneficial as it:
  - Reduces redundancy
  - Aligns better with existing class attributes

- **Final Implementation**:
  ```python
  {"run_name": self.role}
  ```
  This substitution helps maintain consistency and improve the model's semantic clarity.

**Recommendation**: Implement validation for the `role` field to ensure it adheres to required formats, enhancing the quality of data within the class.
```python
@validator('role')
def validate_role(cls, v):
    if not v or not isinstance(v, str):
        raise ValueError("Role must be a non-empty string")
    return v.strip()
```

### 2. Decorator Management (Patches 2 & 3)

The integration of the `@track_agent()` decorator has been pivotal for monitoring agent activities:
```python
from agentops.agent import track_agent

@track_agent()
class Agent(BaseModel):
```
However, its implementation raises concerns:
1. **Error Handling**: There is no handling mechanism if the `agentops` dependency is missing.
2. **Configuration Options**: The lack of toggles for enabling/disabling tracking limits the flexibility of agent usage.

**Recommendations**:
1. **Add Error Handling for Missing Dependency**:
    ```python
    try:
        from agentops.agent import track_agent
    except ImportError:
        def track_agent(func):
            return func  # No-op if not available
    ```

2. **Introduce Configuration for Tracking**:
    ```python
    class Agent(BaseModel):
        enable_tracking: bool = Field(
            default=True,
            description="Enable or disable agent tracking functionality"
        )
    ```

### 3. Modifications to the Execute Task Method

The updates to the `execute_task` method now include:
```python
{"run_name": self.role}
```
This change enhances the context for task execution but could be improved further with type hints and validation.

**Recommendations**:
1. **Implement Type Hints and Validation**:
    ```python
    def execute_task(
        self,
        task_prompt: str,
        run_context: Optional[Dict[str, Any]] = None
    ) -> str:
        run_context = run_context or {}
        run_context.update({
            "run_name": self.role,
            "execution_time": datetime.now().isoformat()
        })
        
        return self.agent_executor.run(
            {
                "input": task_prompt,
                "tool_names": self.agent_executor.tools_names,
                "tools": self.agent_executor.tools_description,
            },
            run_context
        )["output"]
    ```

## General Recommendations
1. **Documentation**: Enhance the docstrings within the modified methods, detailing the role and usage of each field, particularly the interactions between `role` and `run_name`.
   
2. **Testing**: 
   - Introduce unit tests targeting the new tracking functionalities.
   - Develop integration tests involving `agentops` interactions.

3. **Code Organization**: Consider moving tracking into a dedicated module to foster maintainability and clarity.

## Summary of Key Improvements Needed
- Introduce robust error handling for optional dependencies.
- Implement validation for critical fields like `role`.
- Expand documentation to support user clarity.
- Increase test coverage for both unit and integration scenarios.
- Incorporate comprehensive type hints throughout the method definitions.

The proposed changes contribute positively to the overall functionality of the agent implementation but require attention in the areas highlighted above for enhanced clarity, reliability, and maintainability.

**Comment by bhancockio** [comment]
Closed for inactivity.

File: src/crewai/agent.py [hunk 1]
```diff
@@ -232,7 +232,7 @@ def execute_task(
                 "input": task_prompt,
                 "tool_names": self.agent_executor.tools_names,
                 "tools": self.agent_executor.tools_description,
-            }
+            },{"run_name":self.role}
```
Comment (mishra-krishna) [inline src/crewai/agent.py:235]: The idea of adding a `run_name` to the agent executor is great for monitoring purposes. However, the current implementation is not correct. The `run_name` should be part of the input dictionary passed to the `invoke` method. Here is how you can fix it:

```python
self.agent_executor.invoke(
    {
        "input": task_prompt,
        "tool_names": self.agent_executor.tools_names,
        "tools": self.agent_executor.tools_description,
        "run_name": self.role,
    }
)
```

**Review by mishra-krishna — COMMENTED** [review]
The idea of adding a `run_name` to the agent executor is great for monitoring purposes. However, the current implementation is not correct. The `run_name` should be part of the input dictionary passed to the `invoke` method. Here is how you can fix it:

```python
self.agent_executor.invoke(
    {
        "input": task_prompt,
        "tool_names": self.agent_executor.tools_names,
        "tools": self.agent_executor.tools_description,
        "run_name": self.role,
    }
)
```

**Pull request closed without merging**